### ISSUE TITLE
Discount Issue with Checkout

### ISSUE THREAD

**Issue by waynebr76** [jekkos/opensourcepos#1336]
### Issue / Question / Bug

Before submitting an issue please make sure you remove the first section of the template and you tick (add a x between the square brakets) and agree with all the following check boxes:

- [x] Checked the current issues database and no similar issue was already discussed
- [x] Read the README, WHATS_NEW and UPGRADE
- [x] Read the FAQ (https://github.com/jekkos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP has php5-gd, php-intl and sockets installed)
- [x] Reporting an issue of an unmodified OSPOS installation
- [x] PHP version is at least 5.5 and not 7.x
- [x] MySQL version is 5.5 or 5.6 and not 5.7
- [x] Ran any database upgrade scripts (e.g. database/2.4_to_3.0.sql)
- [x] Know the version of OSPOS and git commit hash (check the footer of your OSPOS) and will add to my issue report
- [x] Know the name and version of OS, Web server and MySQL and will add to my issue report

### Installation information
Ubuntu 14.04 LTS, Open Source Point Of Sale Version 3.1.0

### Expected behaviour
When adding a discount of 10% etc the total should be displayed and then payment added and checkout confirmed

### Actual behaviour
When adding a discount the correct amount is shown, payment is added but the Complete button does not appear

### Steps to reproduce the issue
Add item to cart, put 10 in discount button add outstanding cash payment, no button appears to complete

**Comment by odiea** [comment]
If you selected Invoice or Quote from the drop down then you will have to also select a Customer for either of those to work.

**Comment by waynebr76** [comment]
Hi thanks for the quick reply, I am performing this in sale mode

**Comment by odiea** [comment]
I could not duplicate your issue.
![2017-06-08 08_02_16-program manager](https://user-images.githubusercontent.com/10962177/26932547-0221c9f8-4c21-11e7-9823-5421a48239ef.png)

**Comment by waynebr76** [comment]
Hi that's strange. I've tested on 2 pcs and my phone and the add payment button dies not change to complete. I used a fresh download from here

**Comment by odiea** [comment]
All I can suggest is make sure Register says Sale. Have you tried it with a customer name selected?

**Comment by waynebr76** [comment]
Yes I have tried these and still the same issue. Invoice works normally but not sale

**Comment by daN4cat** [comment]
How is the tax calculated? Is that applied to the item or else? The issue might be related to the tax changes....

**Comment by waynebr76** [comment]
All of our tax is included so ospos works out the 20% from the price

**Comment by SteveIreland** [comment]
I'm looking into this nowl

**Comment by SteveIreland** [comment]
I believe I have a fix coming soon.  I found there was more than one issue and I have rewritten the code a bit to make it faster and hopefully cleaner. (relatively speaking) and will be submitting a pull request for this by tomorrow night (as long as there isn't a personal or work crisis that needs to be dealt with).

The fix is simple enough but the testing is a bit slow because I have to test it in so many different configurations and use cases (VAT-non cash rounding, VAT-cash rounding, Sales Tax, Sales by Receipt, Quote, Sales by Invoice, (and Work Order in environment).  There are more twists, but those are the only ones that I can realistically commit to!

**Comment by daN4cat** [comment]
The more features we add the more testing scenarios. I believe the next release after this needs to focus on testing because we are not controlling anymore the situation. Not blaming anybody, it's the effect of the way we are working at the moment.

**Comment by jekkos** [comment]
Yes indeed, maybe there should be some feature freeze or just a focus on testing scenarios.
I tried to plead for this when this tax story began. I knew it was going to bring us a huge stabilization effort. There's just too many ways and we can't currently see whether we introduce some regression in the logic there.

For this kind of functionality I think we could easily add some unit tests to verify.. If not then no one will have enough courage to change sth in that logic in the future as it will be too complex to understand

**Comment by jekkos** [comment]
Might be best to try and  add those tests for tax calculation right now since the knowlegde is still fresh.. but then this will first require us to have a build running on travis (which is problably a task I'll have to tackle)

**Comment by daN4cat** [comment]
@waynebr76 latest master should contain @SteveIreland's fix to this issue, please test and confirm it's all fine.

**Comment by waynebr76** [comment]
Thanks, I have reverted to my previous version and I will have to test once I get chance

**Comment by SteveIreland** [comment]
I've been testing it in a vat tax configuration and can't get it to break.  So hopefully it's a big better.

**Comment by daN4cat** [comment]
Hi @SteveIreland, I was testing with am item costing 7.50, VAT 20% included and applied 10% discount:

![image](https://user-images.githubusercontent.com/636048/27764030-9826aedc-5e87-11e7-83f8-09c407ca3af0.png)

Here the decimals configuration:

![image](https://user-images.githubusercontent.com/636048/27764044-ec892310-5e87-11e7-992b-8aeca027ab98.png)

The issue I see is that the total is correct, however the subtotal is rounded down even if it's half pence, whilst the tax is rounded up even if it's half pence.
Is that done on purpose?

Just wondering and what you thought about it.

**Comment by daN4cat** [comment]
And now checking the Detailed Sales Report I see:

![image](https://user-images.githubusercontent.com/636048/27764096-d5aee1f6-5e88-11e7-9386-78990f8e0fba.png)

the rounding is different in the subtotal

**Comment by daN4cat** [comment]
But reprinting the receipt gives the view from the sale register:

![image](https://user-images.githubusercontent.com/636048/27764108-213b8818-5e89-11e7-9023-5ad097327461.png)

**Comment by SteveIreland** [comment]
I'm not a specialist in VAT tax, so I was trying to avoid changing the tax calculation that was in place in 3.0.2.  I've been using 3.0.2 as my point of reference for tax calculations - except for that annoying issue of not being able to complete the sale which I believe is still "fixed" but it probably wouldn't be a bad idea for me to retest it.

I ran your example in 3.0.2 and come up with a subtotal of 5.62  and a tax amount of 1.13

Since that is what is being calculated in the register for the most recent master branch I believe it is correct.  

As you pointed out the reports are still not 100% and are still on my plate to address.

**Comment by daN4cat** [comment]
Probably it's better to round up the tax as it is at the moment and fix the subtotal in the reports. I'll see if I can find the time today for that.

**Comment by daN4cat** [comment]
@SteveIreland I believe to have fixed the rounding issue resulting in a subtotal or total (example above being one).

However I tested mainly with Tax included and VAT so invite you and @odiea to test the US version of the taxation and check totals are right in reports in various scenarios, and mainly consistent with receipt (print & reprint).

I now need to go for a walk as I spent most of the day cleaning up, upgrading and fixing stuff.
